specs/DataSet.rst

```diff
@@ -0,0 +1,253 @@
+
+DataSet.add_metadata(tag=, info=)
+    Adds metadata to the current DataSet.
+    The metadata is stored under the provided tag.
+    It is an error to add metadata to a completed DataSet.
+
+Writing
+-------
+
+DataSet.add_result(**kwargs)
+    Adds a result to the DataSet.
+    Keyword parameters should have the name of a parameter as the keyword and the value to associate as the value.
+    If there is only one positional parameter and it is a dictionary, then it is interpreted as a map from parameter name to parameter value.
+    It is an error for a value for the same parameter to be specified both using a positional parameter or dictionary parameter and using a keyword,
+    It is an error to provide a value for a key or keyword that is not the name of a parameter in this DataSet.
+    It is an error to add a result to a completed DataSet.
+
+DataSet.add_results(args)
+    Adds a sequence of results to the DataSet.
+    The single argument should be a sequence of dictionaries, where each dictionary provides the values for all of the parameters in that result.
+    See the add_result method for a description of such a dictionary.
+    The order of dictionaries in the sequence will be the same as the order in which they are added to the DataSet.
+    It is an error to add results to a completed DataSet.
+
+DataSet.complete()
+    Marks the DataSet as completed.
+
+Access
+------
+
+DataSet.length
+    This attribute holds the current number of results in the DataSet. 
+
+DataSet.is_empty
+    This attribute will be true if the DataSet is empty (has no results), or false if at least one result has been added to the DataSet.
+    It is equivalent to testing if the length is zero.
+
+DataSet.is_completed
+    This attribute will be true if the DataSet is completed or false if it is in progress.
+
+DataSet.get_data(*params, start=, end=)
+    Returns the values stored in the DataSet for the specified parameters.
+    The values are returned as a list of parallel NumPy arrays, one array per parameter.
+    The data type of each array is based on the data type provided when the DataSet was created.
+    If a parameter is optional and no value was provided for one or more results, the corresponding array entries will be the “null” value for the data type: zero for integers, NaN for floats, “” for strings, None for objects.
+    The parameter list may contain a mix of string parameter names, QCoDeS Parameter objects, and ParamSpec objects.
+    If provided, the start and end parameters select a range of results by result count (index). 
+    Start defaults to 0, and end defaults to the current length.
+    If the range is empty -- that is, if the end is less than or equal to the start – then a list of empty arrays is returned.
+
+DataSet.get_parameters()
+    Returns a list of ParamSpec objects that describe the parameters stored in this DataSet.
+
+DataSet.get_metadata(tag=)
+    Returns metadata for this DataSet.
```

Comment:
Yes, that makes sense.

My basic idea is that anything that is JSON-able should be valid as metadata, and that metadata should be thought of as a big, nested dictionary.

Images feel like something that should be associated with the DataSet through the experiment container or some equivalent container -- although it also seems to me that every plot should include a pointer to the data set somehow. Perhaps each DataSet should have a GUID that uniquely identifies it?